app/integration/runtime/src/main/java/io/syndesis/integration/runtime/util/JsonSimplePredicate.java

```diff
@@ -79,7 +80,20 @@ public JsonSimplePredicate(final String expression, final CamelContext context)
 
     @Override
     public boolean matches(final Exchange exchange) {
-        final Message payload = exchange.getIn();
+        // Clone the exchange and set the JSON message converted to a Map /
+        // List as in message.
+        // The intention is that only this predicate acts on the converted
+        // value, but the original in-message still continues to carry the
+        // same format.
+        // The predicated is supposed to be read only with respect to the
+        // incoming message.
+        final Exchange exchangeForProcessing = ExchangeHelper.createCopy(exchange, true);
```

Comment:
I'm not certain of this: if we're running with stream caching (I think we are), this will work, otherwise it might exhaust the any streams present in the message(s) and fail on next exchange processing. Perhaps a test could prove this?

:duck: seems that we were reading the stream below, so I guess it's OK...

Reply:
@zregvart good point! I added a cache of the input stream and added some tests. Can you please have another look? I pushed the changes to this PR